ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/correlation/CorrelationInfoVisitor.java

```diff
@@ -0,0 +1,161 @@
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.hadoop.hive.ql.optimizer.calcite.correlation;
+
+
+import com.google.common.collect.ImmutableList;
+import org.apache.calcite.rel.RelNode;
+import org.apache.calcite.rel.core.CorrelationId;
+import org.apache.calcite.rex.RexCall;
+import org.apache.calcite.rex.RexCorrelVariable;
+import org.apache.calcite.rex.RexNode;
+import org.apache.calcite.rex.RexOver;
+import org.apache.calcite.rex.RexSubQuery;
+import org.apache.calcite.rex.RexShuttle;
+import org.apache.calcite.sql.SqlKind;
+import org.apache.hadoop.hive.ql.optimizer.calcite.HiveRelShuttleImpl;
+import org.apache.hadoop.hive.ql.optimizer.calcite.reloperators.HiveAggregate;
+import org.apache.hadoop.hive.ql.optimizer.calcite.reloperators.HiveFilter;
+import org.apache.hadoop.hive.ql.optimizer.calcite.reloperators.HiveJoin;
+import org.apache.hadoop.hive.ql.optimizer.calcite.reloperators.HiveProject;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Set;
+import java.util.LinkedHashSet;
+
+/**
+ * CorrelationIdVisitor digs out information from a given RexNode about all the RexSubQuery
+ * RexNodes in the top layer. It does not go into RexSubQuery nodes within a RexSubQuery node.
+ */
+public class CorrelationInfoVisitor extends RexShuttle {
+  // List of CorrelationInfo for all the found RexSubQuery, one HiveCorrelationInfo for each
+  // RexSubQuery
+  private List<HiveCorrelationInfo> correlationInfos = new ArrayList<>();
+
+  // True if there is a "NOT" outside of the subquery
+  private boolean notFlag;
+
+  // Prevent direct instantiation.  Information should be retrieved through the static
+  // getCorrelationInfos method at the bottom of the file.
+  private CorrelationInfoVisitor() {
+  }
+
+  // Private method called by static methoc to retrieve information after RexNode was visited
+  private List<HiveCorrelationInfo> getCorrelationInfos() {
+    return correlationInfos;
+  }
+
+  @Override
+  public RexNode visitCall(RexCall call) {
+    // Record if we see not flag before seeing the subquery.
+    // Can handle NOT(NOT(...)) if  it exists
+    if (call.getOperator().getKind() == SqlKind.NOT) {
+      notFlag = !notFlag;
+    }
+    RexNode returnNode = super.visitCall(call);
+    // reset flag after visit.
+    if (call.getOperator().getKind() == SqlKind.NOT) {
+      notFlag = !notFlag;
+    }
+    return returnNode;
+  }
```

Comment:
Could you please add tests when a not expression has multiple subqueries as operands like
```
explain cbo
select id, int_col, year, month from t_test1 s where not (
  s.int_col in (select count(*) from t_test2 t2 where s.id = t2.id) and 
  s.int_col in (select count(*) from t_test3 t3 where s.id = t.3id)
);
```